testopt/testopt.go

```diff
@@ -0,0 +1,132 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package testopt provides objective functions for testing optimization algorithms.
+package testopt
+
+import (
+	"math"
+	"reflect"
+	"testing"
+
+	"github.com/gonum/floats"
+)
+
+// Minimum represents information about an optimal location of a function.
+type Minimum struct {
+	// X is the location of the minimum. If it is unknown or if the location is
+	// not a single point, X is nil.
+	X []float64
+	// F is the value of the objective function at X.
+	F float64
+	// Global indicates whether the minimum is local or global.
+	Global bool
+}
+
+// Function represents a test objective function.
+type Function interface {
+	F(x []float64) float64
+
+	// Minima returns known minima of the function. If no minima exist or if
+	// their location for the given dimension is unknown, nil is returned.
+	Minima() []Minimum
+}
+
+type Gradient interface {
+	Df(x, grad []float64)
+}
+
+type FunctionGradient interface {
+	FDf(x, grad []float64) float64
+}
+
+const (
+	defaultTol     = 1e-8
+	defaultGradTol = 1e-4
+)
+
+func testFunction(f Function, tests []funcTest, t *testing.T) {
+	for i, test := range tests {
+		F := f.F(test.x)
+		if !floats.EqualWithinAbs(F, test.wantF, defaultTol) {
+			t.Errorf("Test %d: %v function value given by F() is incorrect. Want: %v, Got: %v",
+				i, reflect.TypeOf(f), test.wantF, F)
+		}
+
+		if test.wantGrad != nil {
+			if fGradient, isGradient := f.(Gradient); isGradient {
+				grad := make([]float64, len(test.wantGrad))
+				fGradient.Df(test.x, grad)
+
+				dist := floats.Distance(grad, test.wantGrad, math.Inf(1))
+				if dist > defaultGradTol {
+					t.Errorf("Test %d: %v gradient given by Df() is incorrect. |grad - wantGrad|_∞ = %v",
+						i, reflect.TypeOf(f), dist)
+				}
+			}
+
+			fFunctionGradient, isFunctionGradient := f.(FunctionGradient)
+			if isFunctionGradient {
+				grad := make([]float64, len(test.wantGrad))
+				F := fFunctionGradient.FDf(test.x, grad)
+
+				if !floats.EqualWithinAbs(F, test.wantF, defaultTol) {
+					t.Errorf("Test %d: %v function value given by FDf() is incorrect. Want: %v, Got: %v",
+						i, reflect.TypeOf(f), test.wantF, F)
+				}
+
+				dist := floats.Distance(grad, test.wantGrad, math.Inf(1))
+				if dist > defaultGradTol {
+					t.Errorf("Test %d: %v gradient given by FDf() is incorrect. |grad - wantGrad|_∞ = %v",
+						i, reflect.TypeOf(f), dist)
+				}
+			}
+		}
+	}
+
+	if minima := f.Minima(); minima != nil {
+		for i := range minima {
+			if minima[i].X == nil {
```

Comment:
It's not true for ExtendedRosenbrock. There are multiple local minima in the higher dimensions. In any event, the result is the same as far as an optimizer is concerned -- the true test for finding a minima is finding a point with a sufficiently small gradient. The function value tests are only useful for confirming you have implemented the problem correctly.

With that in mind, if we have implemented the function properly and have found a point with zero gradient, then we have found an unlisted minimum.